packages/medusa/src/api/store/carts/[id]/complete/route.ts

```diff
@@ -1,22 +1,48 @@
 import { completeCartWorkflow } from "@medusajs/core-flows"
-import { MedusaError } from "@medusajs/utils"
+import {
+  MedusaError,
+  ModuleRegistrationName,
+  TransactionState,
+} from "@medusajs/utils"
 import { MedusaRequest, MedusaResponse } from "../../../../../types/routing"
+import { prepareRetrieveQuery } from "../../../../../utils/get-query-config"
+import { refetchEntity } from "../../../../utils/refetch-entity"
 import { refetchOrder } from "../../../orders/helpers"
 import { refetchCart } from "../../helpers"
 import { defaultStoreCartFields } from "../../query-config"
 import { StoreCompleteCartType } from "../../validators"
-import { prepareRetrieveQuery } from "../../../../../utils/get-query-config"
 
 export const POST = async (
   req: MedusaRequest<StoreCompleteCartType>,
   res: MedusaResponse
 ) => {
   const cart_id = req.params.id
 
+  const eventBusModule = req.scope.resolve(ModuleRegistrationName.EVENT_BUS)
+
   const { errors, result } = await completeCartWorkflow(req.scope).run({
     input: { id: cart_id },
     context: { transactionId: cart_id },
     throwOnError: false,
+    events: {
+      onFinish: async ({ transaction }) => {
+        const state = transaction.getState()
+
+        // We only want to emit the order.placed event if the workflow completed
+        if (state !== TransactionState.DONE) {
```

Comment:
A general question - should the event emission be part of the workflow? I.e. if we fail to emit the order placed event, do we want to rollback?

I think if another critical system depends on these events, that might be the desired behavior, so having reliable events is probably quite important

Reply:
+1

Reply:
Hmm, I did think of this. 

There are many ways of integrating events into workflows, and I think we will need to differentiate the types of events we want, to know where something fits well. 

The types of events I can think of right now are:
- Pure module events: Events emitted within modules when entities are created, updated, or deleted, and there can be A LOT of these during the lifetime of a workflow execution. These will need to be grouped somehow and emitted at the very end.
- Workflow lifecycle events: Events that are emitted when workflow-specific state changes, e.g. step succeeds, workflow reverts, etc.
- Domain-specific events: Events that are human-readable and specific to the context in which the workflow is executed, e.g. completing a cart (which is what we do here). 

There are likely more types of events than this.

Maybe we have to dictate what guarantees the different types of events give? For example, I am not sure I want to roll back the entire cart-completion because `order.placed` failed to be emitted. Maybe you should use `order.created` if you want to guarantee.

-----------------

Separately, if we choose to incorporate the event as a step, it **must** be the last step, which is something we currently can't guarantee because workflows can be extended anywhere in the flow.

Reply:
Whether you want to rollback or not will depend on how you use events. If it is just for emails, then sure, you don't want to rollback. However, if you have another system you want to keep in sync (eg. reporting taxes to the tax office), having these events unreliable will be a big issue.

I agree with you, we need to specify what sort of guarantees we offer for events and let people build their systems around them. However, I think that is orthogonal to where the events are emitted, because you can have an optional workflow step that is sending an event, and if it fails we can still complete the workflow, which might be a better solution in this case. WDYT?

As a side note, if users have an overview of all the steps that have failed, (and ideally they can retry), they can emit the events manually and still ensure eventual consistency.

Reply:
> Whether you want to rollback or not will depend on how you use events. If it is just for emails, then sure, you don't want to rollback. However, if you have another system you want to keep in sync (eg. reporting taxes to the tax office), having these events unreliable will be a big issue.

Yes, exactly – I agree, which is why I wanted the differentiation between these events.

> you can have an optional workflow step that is sending an event, and if it fails we can still complete the workflow, which might be a better solution in this case

I am fine going with this for now. As I mentioned, the main reason for choosing the `onFinish` over a step right now is that we can't control the extensions to a workflow, and I wanted to be guaranteed this was emitted at the end.

Lemme try to change it to use a step and I'll ask for a re-review from everyone.

Reply:
@olivermrbl I think it's actually fine even if it gets extended right? If someone extends it to do something else after an order is placed, they can emit another event that they can react to. 

Semantically speaking, after this workflow is done, an order has been placed. Now, if you want extend the workflow to print a **thank you** note, you can have your workflow send a `order.thank_you_printed` (or whatever 😄) and you can listen to that instead/as well.

Now if you want to inject something before the `order.placed` event is triggered, then we'll need these injection/extension zones we've talked about

Reply:
@olivermrbl if you pass in the group_id to the event in the steps/workflow, it will only be emitted once the workflow is completed.

Reply:
Good point, its on step 3 of the events project (we're currently at 1).

Reply:
this is not implemented yet. but this will be the behavior and probably the priority for that will be increased.

Reply:
> its on step 3 of the events project

@riqwan where do you see these steps? I can't find the project

Reply:
on the events RFC, sending you the link